source/common/secret/secret_manager_impl.h

```diff
@@ -1,22 +1,39 @@
 #pragma once
 
+#include <shared_mutex>
 #include <unordered_map>
 
 #include "envoy/secret/secret_manager.h"
+#include "envoy/server/instance.h"
 #include "envoy/ssl/tls_certificate_config.h"
 
 #include "common/common/logger.h"
+#include "common/secret/sds_api.h"
 
 namespace Envoy {
 namespace Secret {
 
 class SecretManagerImpl : public SecretManager, Logger::Loggable<Logger::Id::upstream> {
 public:
-  void addOrUpdateSecret(const envoy::api::v2::auth::Secret& secret) override;
-  const Ssl::TlsCertificateConfig* findTlsCertificate(const std::string& name) const override;
+  SecretManagerImpl(Server::Instance& server) : server_(server) {}
+
+  void addStaticSecret(const envoy::api::v2::auth::Secret& secret) override;
+  const Ssl::TlsCertificateConfig* findStaticTlsCertificate(const std::string& name) const override;
+
+  DynamicSecretProviderSharedPtr
+  findOrCreateDynamicSecretProvider(const envoy::api::v2::core::ConfigSource& config_source,
+                                    std::string config_name) override;
 
 private:
-  std::unordered_map<std::string, Ssl::TlsCertificateConfigPtr> tls_certificate_secrets_;
+  Server::Instance& server_;
+
+  // Manages pairs of secret name and Ssl::TlsCertificateConfig.
+  std::unordered_map<std::string, Ssl::TlsCertificateConfigPtr> static_tls_certificate_secrets_;
+  mutable std::shared_timed_mutex static_tls_certificate_secrets_mutex_;
+
+  // map hash code of SDS config source and SdsApi object.
+  std::unordered_map<std::string, std::weak_ptr<DynamicSecretProvider>> dynamic_secret_providers_;
```

Comment:
Where do you handle removal? If there is no removal do we need a weak pointer here? Though it seems like we should handle removal?

Reply:
From the high level design PR: https://github.com/envoyproxy/envoy/pull/3748
Or a smaller PR (https://github.com/istio/envoy/pull/4) which has removed the changes from this PR and https://github.com/envoyproxy/envoy/pull/3754

Here is the lifetime and owenship of DynamicSecretProvider.

1) Same config_source shares same DynamicSecretProvider object
2) Each ContextConfigImpl owns a ref_count to the DynamicSecretProvider object.
3) ContextConfigImpl is owned by SslSocketFactory which is owned by ListenImpl.

When all ListenImpl instances are going away, DynamicSecretProvider object will be deleted.

The relationship between SecretCallback and DynamicSecretProvider.
1) Multiple ContextConfigImpl instances may share same DynamicSecretProvider If they have same config_source.
2) SslSocketFactory owns ContextConfigImpl and Context.  It implements SecretCallback interface to wait for dynamic secret update.  The callback is registered to DynamicSecretProvider. When the secret is updated, the callback will be called.  SslSocketFactory will create a new Context, the new connections will start to use the new Context.
3) Context is shared_ptr.  The ssl_socket is holding a ref_count of Context to keep the Context it is using alive.

Reply:
I have updated the PR description, that covers the lifetime and ownership of DynamicTlsCertificateSecretProvider. This DynamicTlsCertificateSecretProvider object will be removed if nobody is using it. The weak_ptr<DynamicTlsCertificateSecretProvider> entry will remain in this map.